### ISSUE TITLE
Lets for pure values for monad syntax

### ISSUE THREAD

**Issue by AndreasLoow** [HOL-Theorem-Prover/HOL#661]
Would it be possible to add a way to bind pure values in HOL's monad syntax in the same way one can do in Haskell's monad syntax? For example, something such that the following

```
compile_exp n (BUOp Not e) =
 let (n, tmpvar) = compile_new_name n in do (* bad order... *)
  (n, inp, cs) <- compile_exp n e;
  return (n, VarInp tmpvar, SNOC (Cell1 CNot tmpvar inp) cs)
od
```

could be written as

```
compile_exp n (BUOp Not e) = do
 let (n, tmpvar) = compile_new_name n; (* bad order... *)
 (n, inp, cs) <- compile_exp n e;
 return (n, VarInp tmpvar, SNOC (Cell1 CNot tmpvar inp) cs)
od
```

**Comment by mn200** [comment]
I'd like something in this vein myself, but unfortunately the parser is not clever enough to allow for reuse of the `let` keyword.  Is there some variant you can think of that would still be preferable to having the `let` on the outside?

**Comment by AndreasLoow** [comment]
`mlet` maybe? That is,

```
compile_exp n (BUOp Not e) = do
 mlet (n, tmpvar) = compile_new_name n; (* bad order... *)
 (n, inp, cs) <- compile_exp n e;
 return (n, VarInp tmpvar, SNOC (Cell1 CNot tmpvar inp) cs)
od
```

**Comment by mn200** [comment]
Looks reasonable to me.

**Comment by mn200** [comment]
Should such `let`s compile to a standard HOL `LET` underneath, or should it compile to a bind of a unit?  For example, 

    “let x = e1 in monadic_e2”

*vs*

    “x <- unit e1; monadic_e2”

**Comment by AndreasLoow** [comment]
Hm, what I had in mind was the LET-based translation, but I'm not really sure about the pros and cons between the different alternatives.

**Comment by mn200** [comment]
Closed by da74731
